docs/rfcs/029-sharding-phase1.md

```diff
@@ -0,0 +1,365 @@
+
+**LSN dimension**: this just means the range of LSNs (history), when talking about the range
+of keys and LSNs as a two dimensional space.
+
+## Implementation
+
+### Key sharding vs. LSN sharding
+
+When we think of sharding across the two dimensional key/lsn space, this is an
+opportunity to think about how the two dimensions differ:
+- Sharding the key space distributes the _write_ workload of ingesting data
+  and compacting.  This work must be carefully managed so that exactly one
+  node owns a given key.
+- Sharding the LSN space distributes the _historical read_ workload.  This work
+  can be done by anyone without any special coordination, as long as they can
+  see the remote index and layers.
+
+The key sharding is the harder part, and also the more urgent one, to support larger
+capacity databases.  Because distributing historical LSN read work is a relatively
+simpler problem that most users don't have, we defer it to future work.  It is anticipated
+that some quite simple P2P offload model will enable distributing work for historical
+reads: a node which is low on space can call out to peer to ask it to download and
+serve reads from a historical layer.
+
+### Key mapping scheme
+
+Having decided to focus on key sharding, we must next decide how we will map
+keys to shards.  It is proposed to use a "wide striping" approach, to obtain a good compromise
+between data locality and avoiding entire large relations mapping to the same shard.
+
+We will define three spaces:
+- Key space: positive integer
+- Stripe space: positive integer
+- Shard space: integer from 0 to N-1, where we have N shards.
+
+### Stripe -> Shard mapping
+
+The main property want want from our stripe->shard mapping is
+to spread bandwidth when high throughput contiguous writes are
+happening, such as appending to a relation.
+
+Stripes are mapped to shards with a simple modulo operation:
+
+```
+Stripe | 00 | 01 | 02 | 03 | 04 | 05 | 06 | 07 | 08 | 09 | ...
+Shard  | 00 | 01 | 02 | 01 | 02 | 03 | 01 | 02 | 03 | 01 | ...
+```
+
+### Key -> Stripe mapping
+
+Keys are currently defined in the pageserver's getpage@lsn interface as follows:
+```
+pub struct Key {
+    pub field1: u8,
+    pub field2: u32,
+    pub field3: u32,
+    pub field4: u32,
+    pub field5: u8,
+    pub field6: u32,
+}
+
+
+fn rel_block_to_key(rel: RelTag, blknum: BlockNumber) -> Key {
+    Key {
+        field1: 0x00,
+        field2: rel.spcnode,
+        field3: rel.dbnode,
+        field4: rel.relnode,
+        field5: rel.forknum,
+        field6: blknum,
+    }
+}
+```
+
+_Note: keys for relation metadata are ignored here, as this data will be mirrored to all
+shards.  For distribution purposes, we only care about user data keys_
+
+The properties we want from our Key->Stripe mapping are:
+- Locality in `blknum`, such that a contiguous range of `blknum` will generally fit
+  into the same stripe and consequently land on the same
```

Comment:
Right, that's part of what is proposed here, the language is just a bit sloppy.  Shouldn't have said "generally" since the general case is that different blknum can land on different shard, but the most frequent case is that neighbours will end up on the same shard.